pymc_marketing/mmm/data_generator.py

```diff
@@ -0,0 +1,521 @@
+                suffix="_adstock", fig=fig, ax=ax, title='Adstock Effect', alpha=.5)
+
+
+    def add_saturation_effect(self, lambdas_scale:float=3.5, lambas_sd:float=1):
+        ''' Apply a logistic saturation transformation
+        Parameters
+        ----------
+        lambdas_scale: float
+            Mean of the normal distribution the lambdas are sampled from.
+        lambdas_sd: float
+            Standard deviation of the normal distribution the lambdas are sampled from.
+        ''';
+        self.is_saturated = True
+        # column reference will be "x1_adstock" if adstock has been applied
+        # or just "x1" if adstock was not opted for
+        lambdas = np.random.normal(lambdas_scale, lambas_sd, self.num_channels)
+        self.lambdas = lambdas
+
+        for i, λ in enumerate(lambdas):
+            col_ref = [f"x{i}",f"x{i}_adstock"][self.is_lagged]
+            col_name = col_ref+"_saturated"
+            self.df[col_name] = logistic_saturation(x=self.df[col_ref].to_numpy(), lam=λ).eval()
+
+
+    def plot_saturation_effect(self):
+        '''Plots the raw channels compared to saturation effects are applied.
+        If lagging has been applied, then the lagged signal is plotted in lieu of the raw signal.'''
+        assert self.is_saturated
+        if not self.is_lagged:
+            fig, ax = self.plot_channel_spends(show=False)
+            self.plot_channel_spends(
+                    suffix="_saturated", title="Saturation Effect", fig=fig, ax=ax, alpha=.5)
+        else:
+            fig, ax = self.plot_channel_spends(suffix="_adstock", show=False)
+            self.plot_channel_spends(
+                    suffix="_adstock_saturated", alpha=.5, title="Saturation Effect",
+                    fig=fig, ax=ax)
+
+
+    def add_trend_and_seasonality_effects(self, trend_scale=5, seasonality_scale=3):
+        ''' 3. Trend & Seasonal Components
+        - Now we add synthetic trend and seasonal components to the effect signal.
+        - Only a single degree of Forier seasonality is currently implemented.
+        Parameters
+        ----------
+        trend_scale: float
+            Mean of exponential distribution that trend is sampled from.
+        seasonality_scale: float
+            Mean of exponential distribution that seasonality is sampled from.
+        '''
+        self.is_trended_and_seasonal = True
+        df = self.df
+        self.trend = np.random.exponential(trend_scale)
+        df["trend"] = np.linspace(start=0.0, stop=self.trend, num=self.n)
+
+        # TODO Generalize first order seasonality to Forier degree n
+        (self.sin_coef, self.cos_coef) = np.random.exponential(seasonality_scale, size=2)
+        num_periods = len(df)
+        freq = 52 # weeks per year
+        df["cs"] = -self.sin_coef * np.sin(num_periods * 2 * np.pi * df.index / freq)
+        df["cc"] =  self.cos_coef * np.cos(num_periods * 2 * np.pi * df.index / freq)
+        df["seasonality"] = 0.5 * (df["cs"] + df["cc"])
```

Comment:
This already exists in the mmm/utils.py 
Could that be utilized?

Reply:
That native approach of creating 2N columns and dotting with the sin and cos coefs would make the df unwieldy to pass around. I prefer keeping them compressed into a single 'seasonality' column.

I did however generalize this to N degrees of Fourier series.